plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/CustomInternalTypeSystem.java

```diff
@@ -71,94 +73,66 @@ public Throwable fillInStackTrace() {
+	 *
+	 * Without the call to <code>#propagateIfCancelException()</code> in this override, the cancel exception would be
+	 * suppressed by Xsemantics and the <code>println()</code> would be executed.
+	 */
 	@Override
-	public void checkNotNull(Object object) throws RuleFailedException {
-		if (object == null)
-			throw new RuleFailedExceptionWithoutStacktrace("passed null object to system");
-	}
-
-	@Override
-	@SuppressWarnings("unchecked")
-	public <T> T environmentAccess(RuleEnvironment environment, Object key,
-			Class<? extends T> clazz) throws RuleFailedException {
-		if (environment == null)
-			throw new RuleFailedExceptionWithoutStacktrace("access to null environment");
-		Object value = environment.get(key);
-		if (value == null)
-			throw new RuleFailedExceptionWithoutStacktrace("no mapping in the environment for: "
-					+ stringRep(key));
-		if (clazz.isAssignableFrom(value.getClass()))
-			return (T) value;
-		else
-			throw new RuleFailedExceptionWithoutStacktrace("mapped value " + stringRep(value)
-					+ " cannot be assigned to " + stringRep(clazz));
-	}
-
-	@Override
-	protected RuleFailedException noSuchMethodException(
-			final String judgmentSymbol,
-			final Iterable<String> relationSymbols, Object... params) {
-		return new RuleFailedExceptionWithoutStacktrace("cannot find a rule for "
-				+ judgmentSymbol + " "
-				+ stringRepForParams(params, relationSymbols));
+	public RuleFailedException extractRuleFailedException(Exception e) {
+		operationCanceledManager.propagateIfCancelException(e);
+		return super.extractRuleFailedException(e);
 	}
 
+	/** Overridden to avoid {@link OperationCanceledException}s being suppressed by Xsemantics. */
 	@Override
-	protected RuleFailedException noSuchMethodException(final String name,
-			Object... params) {
-		return new RuleFailedExceptionWithoutStacktrace("cannot find an implementation for "
-				+ name + "(" + stringRepForParams(params) + ")");
+	public void throwRuleFailedException(String message, String issue, Throwable t,
+			ErrorInformation... errorInformations) {
+		operationCanceledManager.propagateIfCancelException(t);
+		super.throwRuleFailedException(message, issue, t, errorInformations);
 	}
 
 	@Override
-	public void throwForExplicitFail(String message,
-			ErrorInformation errorInformation) {
-		final RuleFailedException ex = new RuleFailedExceptionWithoutStacktrace(message);
-		ex.addErrorInformation(errorInformation);
-		Exceptions.sneakyThrow(ex);
+	protected RuleFailedException createRuleFailedException(String message, String issue, Throwable t) {
+		return new RuleFailedExceptionWithoutStacktrace(message, issue, t);
 	}
 
+	/** Overridden to avoid excessive string conversions. */
 	@Override
-	protected RuleFailedException createRuleFailedException(String message, String issue, Throwable t) {
-		return new RuleFailedExceptionWithoutStacktrace(message, issue, t);
+	protected String stringRepForEnv(RuleEnvironment ruleEnvironment) {
+		return "[...]";
```

Comment:
Note: this is old code (I only moved it).

It is useful. The idea is to suppress the full-blown string representation, because that led to performance issues and other problems.